### PR TITLE
chainHead/fix: Report bestBlock events only for newBlock reports

### PR DESCRIPTION
**Pull request by lexnv**
The https://github.com/paritytech/polkadot-sdk/issues/5512 has surfaced that we reported a `BestBlock` event for a block not previously reported via `NewBlock`.

This is because of a race between:
- the stream of events that announces new blocks
- `self.client.info().best_block`

It is possible that `client.info()` contains newer information than the information polled from the block stream (that may be lagging).

To mitigate this, instead of relying on the client's info use the last finalized block to emit a new event.

There are two cases when a new best block event is emitted:
- The best block is in the pruned list and is reported immediately
- The best block is not a descendant of the last finalized block 

Closes: https://github.com/paritytech/polkadot-sdk/issues/5512 

Thanks @jsdw and @josepot for helping debug this 🙏 

cc @paritytech/subxt-team

### CODE REVIEW

**Comment by josepot** [comment]
First of all, thank you for your quick response and for addressing this so promptly!

I’d like to clarify a few points regarding the generation of multiple `BestBlock` events for the same block after this PR:

> After this PR, multiple `BestBlock` events may be generated for the same block. The RPC-v2 spec does not mention that the `BestBlock` must be a descendant of the previously generated `BestBlock`, because it can simply be on a different fork. Therefore, it is ok to generate multiple `BestBlock` events for the same block.  
> However, the spec mentions that a `BestBlock` event must be generated after a `NewBlock` event.

I’m not entirely sure I fully understood your intent here, but I have some concerns. It's important to note that the actual name of the event is `bestBlockChanged`. In my opinion, this event should only be emitted multiple times if there is a valid reason -specifically, if the best block has genuinely changed, such as due to a switch between different branches.

While the spec doesn't explicitly prohibit emitting the same event multiple times for an unchanged best block, I believe it's not in the spirit of the spec to do so. The event’s name, `bestBlockChanged`, suggests that it should only be triggered when there is an actual change. Emitting redundant or noisy events when the best block hasn’t changed could be misleading and I think it's something we should avoid.

That said, I believe this behavior wouldn’t break anything in PAPI, but it could break things for other libraries, as the spec implies that the event is meant to signal an actual change in the best block:

> The `bestBlockChanged` event indicates that the given block is now considered to be the best block.

> `bestBlockHash` is guaranteed to be equal either to the current finalized block hash, or to a block reported in an earlier `newBlock` event.

Also, the spec states:

> This list of notifications makes it very easy for a JSON-RPC client to follow just the best block updates (listening to just `bestBlockChanged` events).

Given these points, it seems reasonable to expect that `bestBlockChanged` events are only emitted when there is a meaningful change. Generating this event without a genuine change could potentially create unnecessary noise for clients monitoring these updates.

cc @tomaka

File: substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs [hunk 1]
```diff
@@ -552,37 +554,34 @@ where
...
 			// In those cases, we emit a new best block.
+			//
+			// When [`Self::generate_finalized_events`] reports the last finalized block as the best
+			// block, it also sets the `self.current_best_block`, if and only if the block
+			// was never reported as `NewBlock` before.
+			//
+			// Therefore, this condition is equivalent with
+			// `!std::matches(FollowEvent::BestBlockChanged(_), events.last())`.
+			// And that means we know for sure that at least a `NewBlock` was generated
+			// for the last finalized block.
```
Comment (bkchr) [inline substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs:573]: How do we know? All of the blocks could already have been known. Then there will be no `NewBlock` generated?

File: substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs [hunk 2]
```diff
@@ -552,37 +554,34 @@ where
...
 			let is_not_last_finalized = current_best_block != last_finalized;
 
 			if is_in_pruned_list || is_not_last_finalized {
```
Comment (bkchr) [inline substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs:576]: I don't get why you want to generate the `BestBlockChangedEvent` when the best block is not the last finalized? The best block could just be an descendant of the last finalized?

**Comment by tomaka** [comment]
> It's important to note that the actual name of the event is bestBlockChanged. In my opinion, the exact same event should only be emitted multiple times if there is a valid reason -specifically, if the best block has genuinely changed, such as due to a switch between different branches.

I don't think we should focus too much on the name. All APIs always have small prints that aren't conveyed by the name of the thing itself.
Yes, emitting multiple identical events in a row contradicts the word "changed" and technically it is misleading, but I really don't think that this "misleading-ness" a big deal.

Whether it is allowed to emit multiple identical events in a row should be explicitly written in the spec.
Given that the best block can change at any time and in unpredictable ways, I can't manage to find a situation where it would be troublesome for the JSON-RPC client if the best block changed to what it already is, and so it should IMO simply be allowed but slightly discouraged.

File: substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs [hunk 3]
```diff
@@ -454,34 +457,33 @@ where
...
-					}
-				}
+				continue;
+			}
 
-				// Let's generate the `NewBlock` and `NewBestBlock` events for the block.
-				events.extend(self.generate_import_events(*hash, *parent, true))
+			// If we end up here and the `best_block` is a descendent of the finalized block
+			// (last block in the list), it means that there were skipped notifications.
+			// Otherwise `pin_block` would had returned `true`.
```
[the remark this answers is not in the document]
Comment (lexnv) [reply substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs]: Have changed this, thanks!

File: substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs (hunk 1 above)
Comment (lexnv) [reply substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs]: Have removed this comment and added a simpler code path using lowest common ancestor, thanks!

File: substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs [hunk 4]
```diff
@@ -454,34 +457,33 @@ where
...
-							"A descendent of the finalized block was already reported".into(),
-						))
-					}
-				}
+				continue;
+			}
 
-				// Let's generate the `NewBlock` and `NewBestBlock` events for the block.
-				events.extend(self.generate_import_events(*hash, *parent, true))
+			// If we end up here and the `best_block` is a descendent of the finalized block
```
[the remark this answers is not in the document]
Comment (lexnv) [reply substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs]: Have moved the comment next to the lowest common ancestor block, thanks!

File: substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs (hunk 2 above)
Comment (lexnv) [reply substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs]: Yep that makes sense, have checked instead that the best block is a descendant of the last finalized block 🙏

**Comment by lexnv** [comment]
Have modified the code logic to take care of the following edge-cases:
- The best block is in the pruned list and is reported immediately
- The best block is not a descendant of the last finalized block

We now should not emit extra `BestBlock` events if the current best block is a descendant of the finalized block. Thanks for the review everyone 🙏

File: substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs [hunk 5]
```diff
@@ -549,39 +551,36 @@ where
...
+				// Case 2. Check if the best block is a descendant of the last finalized block.
+				let ancestor = sp_blockchain::lowest_common_ancestor(
+					&*self.client,
+					last_finalized,
+					current_best_block,
+				)?;
+
+				let is_descendant = ancestor.hash == last_finalized;
+				// If the best block is not a descendant of the last finalized block, it means that
+				// the best block is on a fork, that will be pruned in the future.
```
Comment (skunert) [inline substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs:578]: Is this comment taking into account that the pruning logic changed? We should now prune directly when finalization takes place and not anymore when the head of a fork is below the finalized block. If we arrive in this codepath maybe the stored best block here is just below the finalized one? Logic still looks correct though.

File: substrate/client/rpc-spec-v2/src/chain_head/tests.rs [hunk 6]
```diff
@@ -3833,3 +3833,222 @@ async fn follow_unique_pruned_blocks() {
...
+	//
+	// finalized -> block 1 -> block 2
+	//                         ^^^ best block reported
+	//
+	//           -> block 1 -> block 2_f -> block 4 -> block 5 (best)
+	//                                    ^^^ finalized
+	//
+	// The block 4 is needed on the longest chain because we want the
+	// best block 2 to be reported as pruned. Pruning is happening at
+	// height (N - 1), where N is the finalized block number.
```
Comment (skunert) [inline substrate/client/rpc-spec-v2/src/chain_head/tests.rs:3881]: Should also not be true anymore after https://github.com/paritytech/polkadot-sdk/pull/3962

File: substrate/client/rpc-spec-v2/src/chain_head/tests.rs [hunk 7]
```diff
@@ -3833,3 +3833,222 @@ async fn follow_unique_pruned_blocks() {
...
+	let block_4_hash = block_4.hash();
+	client.import(BlockOrigin::Own, block_4.clone()).await.unwrap();
+
+	// Import block 2 as best on the fork.
+	let mut block_builder = BlockBuilderBuilder::new(&*client)
+		.on_parent_block(block_1_hash)
+		.with_parent_block_number(1)
+		.build()
+		.unwrap();
+	// This push is required as otherwise block 3 has the same hash as block 2 and won't get
```
Comment (skunert) [inline substrate/client/rpc-spec-v2/src/chain_head/tests.rs:3922]: Block 3 went missing

File: prdoc/pr_5527.prdoc [hunk 8]
```diff
@@ -0,0 +1,17 @@
+# Schema: Polkadot SDK PRDoc Schema (prdoc) v1.0.0
+# See doc at https://raw.githubusercontent.com/paritytech/polkadot-sdk/master/prdoc/schema_user.json
+
+title: Report bestBlock events only for newBlock reports
+
+doc:
+  - audience: Node Dev
+    description: |
+      This PR ensures that the chainHead_v1_follow method of the RPC-v2 API are always
```
Comment (skunert) [inline prdoc/pr_5527.prdoc:9]:
```suggestion
      This PR ensures that the chainHead_v1_follow method of the RPC-v2 API is always
```

File: prdoc/pr_5527.prdoc [hunk 9]
```diff
@@ -0,0 +1,17 @@
+# Schema: Polkadot SDK PRDoc Schema (prdoc) v1.0.0
+# See doc at https://raw.githubusercontent.com/paritytech/polkadot-sdk/master/prdoc/schema_user.json
+
+title: Report bestBlock events only for newBlock reports
```
Comment (skunert) [inline prdoc/pr_5527.prdoc:4]:
```suggestion
title: Report BestBlock events only for newBlock reports
```

File: substrate/client/rpc-spec-v2/src/chain_head/tests.rs [hunk 10]
```diff
@@ -3833,3 +3833,222 @@ async fn follow_unique_pruned_blocks() {
 	});
 	assert_eq!(event, expected);
 }
+
+#[tokio::test]
+async fn follow_report_best_block_of_a_known_block() {
```
Comment (skunert) [inline substrate/client/rpc-spec-v2/src/chain_head/tests.rs:3838]: Are we sure this is testing what was changed here? I wanted to understand this better and ran the test on master to see it fail, but it passed there too 😬

File: substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs (hunk 5 above)
Comment (lexnv) [reply substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs]: That makes sense, wasn't aware the pruning logic changed! I've changed the comment to reflect that, thanks!

File: substrate/client/rpc-spec-v2/src/chain_head/tests.rs (hunk 6 above)
Comment (lexnv) [reply substrate/client/rpc-spec-v2/src/chain_head/tests.rs]: Yep, double checked and adjusted the test, it seems we are now pruning all blocks

File: substrate/client/rpc-spec-v2/src/chain_head/tests.rs (hunk 7 above)
Comment (lexnv) [reply substrate/client/rpc-spec-v2/src/chain_head/tests.rs:3907]: Thanks, have simplified the test a bit and properly named the blocks

File: substrate/client/rpc-spec-v2/src/chain_head/tests.rs (hunk 10 above)
Comment (lexnv) [reply substrate/client/rpc-spec-v2/src/chain_head/tests.rs:3838]: Yep that's correct. Initially, I wanted to check by this test the fact assumption that we can generate multiple BestBlock events (even for blocks that are descendant of the current best block), similar to:
` Best 4, Best 5, Best 4, Finalized 4`.

It turned out to not be as complicated as I thought initially, and have modified the testing to use the internal client mock of chainHead. The test should now cover the exact race condition of `self.client.info()` and fail on origin/master :D Thanks for pointing this out 🙏

**Review by skunert — APPROVED** [review]

**Review by pkhry — APPROVED** [review]

**Comment by niklasad1** [comment]
bot fmt

File: substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs [hunk 11]
```diff
@@ -549,39 +551,32 @@ where
...
+					best_block_hash: last_finalized,
+				}));
+			} else {
+				// The pruning logic ensures that when the finalized block is announced,
+				// all blocks on forks that have the common ancestor lower or equal
+				// to the finalized block are reported.
+				//
+				// However, we double check if the best block is a descendant of the last finalized
+				// block to ensure we don't miss any events.
+				let ancestor = sp_blockchain::lowest_common_ancestor(
```
Comment (bkchr) [inline substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs:569]: I don't get why we need this. If the best block was not pruned, it needs to be a descendant of the finalized block.

**Pull request merged**